crates/transpiler/src/passes/convert_discrete_rotations.rs

```diff
@@ -0,0 +1,369 @@
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use std::f64::consts::PI;
+
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::operations::{OperationRef, Param, StandardGate};
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+const ROTATION_GATE_NAMES: &[&str; 3] = &["rx", "ry", "rz"];
+
+const PI8: f64 = PI / 8.0;
+const MINIMUM_TOL: f64 = 1e-12;
+
+/// For an angle, if it is a multiple of PI/8, calculate the multiplicity mod 16,
+/// Otherwise, return None.
+fn is_angle_close_to_multiple_of_2pi_8(angle: f64, tol: f64) -> Option<usize> {
+    let closest_ratio = angle * 4.0 / PI;
+    let closest_integer = closest_ratio.round();
+    let closest_angle = closest_integer * PI / 4.0;
+    let theta = angle - closest_angle;
+
+    let tr_over_dim = (theta / 2.).cos();
+    let dim = 2.;
+
+    // fidelity-based tolerance
+    let f_pro = tr_over_dim.abs().powi(2);
+    let gate_fidelity = (dim * f_pro + 1.) / (dim + 1.);
+    if (1. - gate_fidelity).abs() < tol {
+        Some((closest_integer as usize) % 16)
+    } else {
+        None
+    }
+}
+
+/// Gets a rotation gate (RX/RY/RZ) and outputs an equivalent vector of standard gates and
+/// a global phase, when the gate is sufficiently close to Clifford+T/Tdg.
+/// Otherwise, return None.
+fn try_replace_rotation_by_discrete(
+    gate: StandardGate,
+    angle: f64,
+    tol: f64,
+) -> Option<(Vec<StandardGate>, f64)> {
+    let multiple = is_angle_close_to_multiple_of_2pi_8(angle, tol);
+    let mut discrete_sequence = Vec::<StandardGate>::with_capacity(4);
+
+    let global_phase = match (gate, multiple) {
+        (StandardGate::RZ, Some(0)) => {
+            discrete_sequence.push(StandardGate::I);
+            0.0
+        }
+        (StandardGate::RZ, Some(1)) => {
+            discrete_sequence.push(StandardGate::T);
+            -1.0 * PI8
+        }
+        (StandardGate::RZ, Some(2)) => {
+            discrete_sequence.push(StandardGate::S);
+            -2.0 * PI8
+        }
+        (StandardGate::RZ, Some(3)) => {
+            discrete_sequence.push(StandardGate::S);
+            discrete_sequence.push(StandardGate::T);
+            -3.0 * PI8
+        }
+        (StandardGate::RZ, Some(4)) => {
+            discrete_sequence.push(StandardGate::Z);
+            -4.0 * PI8
+        }
+        (StandardGate::RZ, Some(5)) => {
+            discrete_sequence.push(StandardGate::Z);
+            discrete_sequence.push(StandardGate::T);
+            -5.0 * PI8
+        }
+        (StandardGate::RZ, Some(6)) => {
+            discrete_sequence.push(StandardGate::Sdg);
+            -6.0 * PI8
+        }
+        (StandardGate::RZ, Some(7)) => {
+            discrete_sequence.push(StandardGate::Tdg);
+            -7.0 * PI8
+        }
+        (StandardGate::RZ, Some(8)) => {
+            discrete_sequence.push(StandardGate::I);
+            -1.0 * PI
```

Comment:
This should also work, right?

```suggestion
            -PI
```

Reply:
done in [066ded3](https://github.com/Qiskit/qiskit/pull/15321/commits/066ded3f9e884d07037336a87d969f43cb54c8eb)